internal/provider/file/resources.go

```diff
@@ -0,0 +1,335 @@
+// Copyright Envoy Gateway Authors
+// SPDX-License-Identifier: Apache-2.0
+// The full text of the Apache license is available in the LICENSE file at
+// the root of the repo.
+
+package file
+
+import (
+	"fmt"
+	"os"
+	"path/filepath"
+	"reflect"
+	"strings"
+
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/util/sets"
+	gwapiv1 "sigs.k8s.io/gateway-api/apis/v1"
+	gwapiv1a2 "sigs.k8s.io/gateway-api/apis/v1alpha2"
+	"sigs.k8s.io/yaml"
+
+	egv1a1 "github.com/envoyproxy/gateway/api/v1alpha1"
+	"github.com/envoyproxy/gateway/internal/envoygateway"
+	"github.com/envoyproxy/gateway/internal/envoygateway/config"
+	"github.com/envoyproxy/gateway/internal/gatewayapi"
+)
+
+// loadFromFilesAndDirs loads resources from specific files and directories.
+func loadFromFilesAndDirs(files, dirs []string) ([]*gatewayapi.Resources, error) {
+	var rs []*gatewayapi.Resources
+
+	for _, file := range files {
+		r, err := loadFromFile(file)
+		if err != nil {
+			return nil, err
+		}
+		rs = append(rs, r)
+	}
+
+	for _, dir := range dirs {
+		r, err := loadFromDir(dir)
+		if err != nil {
+			return nil, err
+		}
+		rs = append(rs, r...)
+	}
+
+	return rs, nil
+}
+
+// loadFromFile loads resources from a specific file.
+func loadFromFile(path string) (*gatewayapi.Resources, error) {
+	if _, err := os.Stat(path); err != nil {
+		if os.IsNotExist(err) {
+			return nil, fmt.Errorf("file %s is not exist", path)
+		}
+		return nil, err
+	}
+
+	bytes, err := os.ReadFile(path)
+	if err != nil {
+		return nil, err
+	}
+
+	return convertKubernetesYAMLToResources(string(bytes))
+}
+
+// loadFromDir loads resources from all the files under a specific directory excluding subdirectories.
+func loadFromDir(path string) ([]*gatewayapi.Resources, error) {
+	entries, err := os.ReadDir(path)
+	if err != nil {
+		return nil, err
+	}
+
+	var rs []*gatewayapi.Resources
+	for _, entry := range entries {
+		// Ignoring subdirectories and all hidden files and directories.
+		if entry.IsDir() || strings.HasPrefix(entry.Name(), ".") {
+			continue
+		}
+
+		r, err := loadFromFile(filepath.Join(path, entry.Name()))
+		if err != nil {
+			return nil, err
+		}
+
+		rs = append(rs, r)
+	}
+
+	return rs, nil
+}
+
+// TODO(sh2): This function is copied and updated from internal/cmd/egctl/translate.go.
+// This function should be able to process arbitrary number of resources, so we
+// need to come up with a way to extend the GatewayClass and EnvoyProxy field to array
+// instead of single variable in gatewayapi.Resources structure.
+//
+// - This issue is tracked by https://github.com/envoyproxy/gateway/issues/3207
+//
+// convertKubernetesYAMLToResources converts a Kubernetes YAML string into GatewayAPI Resources.
+func convertKubernetesYAMLToResources(str string) (*gatewayapi.Resources, error) {
```

Comment:
should deal with `ReferenceGrant` type resource in this func?

Reply:
Yes, it should. This PR only leverage the basic functionality for file-provider, as described in the comments here, this function will be revisited along with the one in egctl x translate.